ReadMe.md
# ARA Tools

<img src="https://github.com/BaselLaserMouse/ara_tools/wiki/images/RegSchematicAssembled.png" />


### What does it do?
This is a collection of MATLAB functions for handling mouse whole-brain imaging data and associated "sparse" point data such as cell locations, traced neurite trees, or traced projections from a bulk injection. 
This package provides functions for relating these sparse data to the Allen Reference Atlas (ARA):

1. Downsampling full-resolution brain volumes to the ARA voxel size. 
2. Downsampling associated sparse data to the ARA voxel size and exporting of these data to a csv file.
3. Registering the sample brain and sparse data to the ARA: both sample -> ARA and ARA -> sample transformations.
4. Determining the ARA brain area associated with each sparse point.
5. Extracting area outlines from the reference atlas that you can use for plotting. 

### Who is it for?
*ARA Tools* is written in the [Mrsic-Flogel lab](http://mouse.vision) with our data in mind.
In places it might expect the raw data were stitched using [StitchIt](https://github.com/BaselLaserMouse/StitchIt), so you may need to adapt our code for your own work.

### Installation

1. Clone the repository in a Git client like [SmartGit](http://www.syntevo.com/smartgit/) or [SourceTree](https://www.sourcetreeapp.com).
2. You will need some or all of the dependencies listed below. Which ones you need depends on what you're doing. 
e.g. you won't need the registration stuff to plot cells. 
3. You will need a copy of the Allen Atlas and the template brain in MHD format. One is included on [this page](http://mouse.vision/han2017). 
5. Add the contents of the `code` directory to your MATLAB path. 
6. Setup ara_tools properly (see [wiki](https://github.com/SainsburyWellcomeCentre/ara_tools/wiki/Setup))

### Dependencies
ARA tools depends on the following MATLAB packages and requires at least R2016b. 
Clone them and add to your path.
- [MelastiX](https://github.com/raacampbell/matlab_elastix) (if you need to register brains to the ARA)
- [yaml matlab](https://github.com/raacampbell/yamlmatlab)
- [Allen API](https://github.com/BaselLaserMouse/AllenBrainAPI)
- [JSONlab](http://www.mathworks.com/matlabcentral/fileexchange/33381-jsonlab--a-toolbox-to-encode-decode-json-files-in-matlab-octave)
- [settings handler](https://github.com/raacampbell/settings_handler)
- [stitchit](https://github.com/BaselLaserMouse/StitchIt)
- [matlab-tree](https://github.com/raacampbell/matlab-tree) (Only for tracing axon trees)

### Also relevant
- [Han 2018 MATLAB cell tracing code](https://github.com/BaselLaserMouse/HanSingleCell)


### Using the toolbox
See the [Wiki](https://github.com/BaselLaserMouse/ara_tools/wiki) for documentation on using the tools. 


### Related projects

* [ClearMap](https://github.com/ChristophKirst/ClearMap)
* [amap](https://github.com/SainsburyWellcomeCentre/amap-python)


### Further info
* [ARA white paper from the Allen Institute](http://help.brain-map.org/download/attachments/2818171/MouseCCF.pdf)
* [The latest CCF](http://download.alleninstitute.org/informatics-archive/current-release/)


### Major changes
1st April 2020 - Large changes to the directory structure of the downsampled stacks and the input arguments and behavior of `ARAregister`. See changelog. Look in the [releases](https://github.com/SainsburyWellcomeCentre/ara_tools/releases) if you want to just use the latest legacy version. 


Changelog.txt
Changelog

17/08/2021
Allow for all downsampled channels to be transformed in the sample2ARA directory using the calculated
transformation. This is done with this new form of calling ARAregister:
  ARAregister('channel',4,'transformAllChannels',true)
 or - 
  ARAregister(transformAllChannels',true)

You may also transform named downsampled stacks as follows:
>> cd /mnt/data/mySampleDir
>> dsFile='downsampled_stacks/025_micron/ds_XY_sample_25_25_ch02_chan_2_red.tif'
>> regDir='registration/reg_01__2021_08_16_a/sample2ARA/';
>> aratools.applyRegToStackFile(dsFile,regDir)


01/04/2020
MAJOR CHANGES TO ARAregister - 
* Downsampled stacks and registration directory are now separate. There can be multiple registration directories.
The new folder layout for downsampled stacks is produced using the master branch of StitchIt as of 01/04/2020. 
The directory layout of the sample dir is now:

├── recipe_sample_1069012_200313_153543.yml
├── acqLog_sample_1069012.txt
├── downsampled_stacks
│   ├── 025_micron
│   │   ├── ds_sample_1069012_200313_153543_25_25_ch02.tif
│   │   ├── ds_sample_1069012_200313_153543_25_25_ch02.txt
│   │   ├── ds_sample_1069012_200313_153543_25_25_ch03.tif
│   │   └── ds_sample_1069012_200313_153543_25_25_ch03.txt
│   └── 050_micron
│       ├── ds_sample_1069012_200313_153543_50_50_ch02.txt
│       ├── ds_sample_1069012_200313_153543_50_50_ch02_red.tif
│       ├── ds_sample_1069012_200313_153543_50_50_ch03.txt
│       └── ds_sample_1069012_200313_153543_50_50_ch03_green.tif
├── registration
│   ├── reg_01__2020_03_23_a
│   │   ├── registration_log.txt
│   │   ├── ARA_to_sample
│   │   │   ├── ARA_to_sample_parameters_1.txt
│   │   │   ├── ARA_to_sample_parameters_2.txt
│   │   │   ├── CMD
│   │   │   ├── etc...
│   │   │
│   │   └── sample_to_ARA
│   │       ├── CMD
│   │       ├── IterationInfo.0.R0.txt
│   │       ├── IterationInfo.0.R1.txt
│   │       ├── etc...
│   │
│   └── reg_02__2020_03_23_b
│       ├── ARA_to_sample
│       │   ├── etc...
│       │ 
│       └── registration_log.txt

** You will need to delete your local settings file in ~/.defaultSettings_ARAtools.yml before you run.
You can choose which atlas pixel size to use for registration by editing that file. 

* Please also see help ARAregister to learn more about the new options available. 




3/07/2019
The following changes may break previous behavior
- transformSparsePoints now just treats the last three columns as being point coordinates.
- invertExportedSparseFiles no longer relies on the logging file

27/07/2019
ARA_tools will slowly become more general purpose and may no longer work with older data. 
There is a commit tag to indicate the point beyond which compatibility with Han 2017 is no longer guaranteed. 

28/07/2019
Running "aratools.rescaleAllSparsePoints" in data dir will rescale sparse points CSV files in the format Z,X,Y

code/atlas_utils/ARAregister.m
function ARAregister(varargin)
% Register sample brain to the ARA and create transform parameters for sparse points
%
% function ARAregister('param1',val1, 'param2',val2, ...)
%
% Purpose
% Register sample brain to the ARA (Allen Reference Atlas) in various ways. 
% This function assumes you have a directory containing your brain in the same
% voxel size as the ARA. You can generate this for instance, by running the
% downsampleAllChannels command from stitchit.
%
% By default this function:
% 1. Registers the ARA template TO the sample.
% 2. Registers the sample to the ARA template.
% 3. If (2) was done, the inverse transform of it is also calculated and applied
%    to both volume images and sparse data.
%
%
% The results are saved to downsampleDir
%
% If no inputs are provided it looks for the default down-sample directory. 
% The ARA to use is infered from the file names in downsampleDir. 
% NOTE: once this function has been run you can transform the sparse points to ARA
%       again simply by running invertExportedSparseFiles from the experiment root dir.
%
%
% Inputs (optional parameter/value pairs)
% 'downsampleDir' - String defining the directory that contains the downsampled data. 
%                   By default uses value derived from the toolbox YML file using 
%                   the function aratools.getDownSampledDir.
% 'ara2sample' - [bool, default true] whether to register the ARA to the sample
% 'sample2ara' - [bool, default true] whether to register the sample to the ARA
% 'suppressInvertSample2ara' - [bool. default false] if true, the inverse transform is not
%                            calculated if the sample2ara transform is performed.
%                            You need the inverse transform if you want to go on to 
%                            register sparse points to the ARA. 
% 'elastixParams' - paths to parameter files. By default we use those in ARA_tools/elastix_params/
% 'channel' - If missing an interactive selector at the command line appears. If present,
%             channelToRegister should be an integer that corresponds channel IDs embdedded 
%             in downsampled file names. For example: dsXYZ001_25_25_ch02.tiff is channel 2.
%             NOTE: this input is ignored if only one downsampled file is present. 
% 'medFiltSample' - False by default. If true, the sample image is filtered with a 3D median
%             filter of width 7 before registration. If medFiltSample is an integer greater 
%             than 1, then the sample image is filtered by this amount instead. 
%             NOTE: The *unfiltered* result image will appear in the registration directory 
%             as "result.tiff".
% 'transformAllChannels' - If true, we run the sample2ARA transform on all available channels 
%            once the registration is done. The output saved as tiffs in sample2ARA. The channel
%            used for registration is the "result" file. False by default.
% 'action' - If a registration already has been done, then ARAregister does nothing. This is 
%            over-ridden by the "action" argument:
%            'delete' -- deletes existing registrations (WITH NO CONFIRMATION) and replace all with a new one. 
%            'keep' -- retain existing registrations but add a new one.
%            'tidy' -- delete all registrations apart from a named one which is chosen at the CLI. Then do nothing.
%
%
% Outputs
% none
%
%
% For more details see the repository ReadMe file and als see the wiki
% (https://github.com/SainsburyWellcomeCentre/ara_tools/wiki/Example-1-basic-registering). 
%
%
% Examples
% - Run with defaults
% >> ARAregister
%
% - Run with another set of parameter files
% >> ARAregister('elastix_params','ParamBSpline.txt'})
%
% - Run on channel 4
% >> ARAregister('channel',4)
%
% - Run registration with a filtered sample stack (this may improve the registration significantly)
% >> ARAregister('medFiltSample',true)
%
% - Register using channel 4 then transform all remaining channels sample2ara channels using the same transform
%  ARAregister('channel',4,'transformAllChannels',true)
%. (to do this manually for a downsampled stack see aratools.applyRegToStackFile)
% 
% Rob Campbell - Basel, 2015
%
% Also see from this repository:
% invertExportedSparseFiles (and transformSparsePoints), aratools.rescaleAllSparsePoints
%
% Changes
% 2020/03/18 - Modifies the parameters automaticaly for the correct voxel size. It does this
%              by loading them into a structure, altering this, then feeding this into the 
%              elastix function.
% 2021/08/17 - Provides for automatic transformation of all remaining channels.


%Parse input arguments
S=settings_handler('settingsFiles_ARAtools.yml');

params = inputParser;
params.CaseSensitive=false;

params.addParamValue('downsampleDir', aratools.getDownSampledDir,@ischar)
params.addParamValue('ara2sample', true, @(x) islogical(x) || x==1 || x==0)
params.addParamValue('sample2ara', true, @(x) islogical(x) || x==1 || x==0)
params.addParamValue('channel', [], @(x) isnumeric(x) && isscalar(x))
params.addParamValue('suppressInvertSample2ara', false, @(x) islogical(x) || x==1 || x==0)
params.addParamValue('medFiltSample', false, @(x) islogical(x) || (isint(x) && x>=0))
params.addParamValue('transformAllChannels', false, @(x) islogical(x) || x==1 || x==0)
params.addParamValue('action', 'none', @(x) ischar(x) && (strcmpi('none',x) || strcmpi('delete',x) || strcmpi('keep',x) || strcmpi('tidy',x)))


toolboxPath = fileparts(which(mfilename));
toolboxPath = fileparts(fileparts(toolboxPath));
elastix_params_default = {fullfile(toolboxPath,'elastix_params','01_ARA_affine.txt'),
                fullfile(toolboxPath,'elastix_params','02_ARA_bspline.txt')};
params.addParamValue('elastixParams',elastix_params_default,@iscell)


params.parse(varargin{:});
downsampleDir = params.Results.downsampleDir;
ara2sample = params.Results.ara2sample;
sample2ara = params.Results.sample2ara;
channel = params.Results.channel;
suppressInvertSample2ara = params.Results.suppressInvertSample2ara;
elastixParams = params.Results.elastixParams;
medFiltSample = params.Results.medFiltSample;
transformAllChannels = params.Results.transformAllChannels;
action = lower(params.Results.action);

if ~exist(downsampleDir,'dir')
    fprintf('%s failed to find downsampled directory "%s"\n', mfilename, downsampleDir), return
end


% Make the registration directory if needed
existingRegDir=aratools.findRegDirs;

if isempty(existingRegDir)
    % If no registration directories exist, we make one
    regDir = aratools.makeRegDir;

elseif ~isempty(existingRegDir) && strcmp('delete',action)
    % Delete existing registration directory tree and build a new one
    if exist(S.regDir,'dir')
        fprintf('Deleting all existing registrations\n')
        rmdir(S.regDir,'s')
    end
    regDir = aratools.makeRegDir;

elseif ~isempty(existingRegDir) && strcmp('keep',action)
    % Add a new registration directory
    regDir = aratools.makeRegDir;
    fprintf('Registering into new directory %s\n', regDir)

elseif ~isempty(existingRegDir) && strcmp('tidy',action)
    % Delete all but one registration
    aratools.tidyRegistrations
    return

else
    % Otherwise we ask the user what to do
    fprintf('\nFound the following existing registration directories:\n')
    cellfun(@(x) fprintf('%s\n',x), existingRegDir)
    fprintf('\nDoing nothing!\nIf you wish, you may re-run ARAregister using the ''action'' parameter to do one of the following:\n')
    fprintf('1. DELETE existing registrations and replace all with a NEW one.\n')
    fprintf('2. KEEP existing registrations but ADD a new one.\n')
    fprintf('3. TIDY by DELETING all registrations apart from a named one.\n')
    return
end


if sample2ara && suppressInvertSample2ara
    invertSample2ara = false;
else
    invertSample2ara = true ;
end

% Set median filter width to a default value if the user set this input argument to true
if medFiltSample==true
    medFiltSize=7;
end

% Handle scenario where the user supplied a width to filter by
if medFiltSample>1
    medFiltSize = medFiltSample;
    medFiltSample = true;
    if mod(medFiltSample,2)==0
        medFiltSize=medFiltSize+1;
    end
end


%Check that the elastixParams are there
for ii=1:length(elastixParams)
    if ~exist(elastixParams{ii},'file')
        error('Can not find elastix param file %s',elastixParams{ii})
    end
end

% We will now load the parameters into a cell array of structures and then modify it
% so that the correct pixel size is used. 
for ii=1:length(elastixParams)
    elastixParams{ii}=elastix_parameter_read(elastixParams{ii});
    if isfield(elastixParams{ii},'FinalGridSpacingInVoxels')
        elastixParams{ii}.FinalGridSpacingInVoxels(:) = S.ARAsize;
    end
end

%Figure out which atlas to use
dsFile = aratools.getDownSampledFile; %This cell array will later be replaced by a string that is the channel to register
nonRegChannels = dsFile; % This cell array will later have the channel to register removed

if isempty(dsFile)
    return %warning message already issued
end


if iscell(dsFile)
    if length(dsFile) == 1
        dsFile = dsFile{1};
    elseif ~isempty(channel)
        % Find the selected channel in the list of file names
        tok=cellfun(@(x) regexp(x,'.+_\d+_\d+_ch(\d+).+','tokens'),dsFile,'UniformOutput',false);
        matchChans = cellfun(@(x) strcmp(x,sprintf('%02d',channel)),[tok{:}]);
        if ~any(matchChans)
            % Requested channel not found
            fprintf('Requested channel %d does not exist. Available files:\n',channel);
            cellfun(@(x) fprintf('%s\n',x), dsFile)
            return
        end
        if sum(matchChans)>1
            % Multiple channels found
            fprintf('Multiple channels match requested channel number %d. Available files:\n',channel);
            cellfun(@(x) fprintf('%s\n',x), dsFile)
            return
        end
        dsFile = dsFile{find(matchChans)};

    else

        %Display choices to screen and allow user to choose which volume to register
        fprintf('\n Which volume do you want to use for registration?\n')
        for ii=1:length(dsFile)
            fprintf('%d. %s\n',ii,dsFile{ii})
        end
        qs=sprintf('[1 .. %d]? ', length(dsFile));
        userAnswer = [];
        while isempty(userAnswer)
            userAnswer = input(qs,'s');
            userAnswer = str2num(userAnswer);
            if ~isempty(userAnswer) && userAnswer>=1 && userAnswer<=length(dsFile)
                break
            else
                userAnswer=[];
            end
        end

        dsFile = dsFile{userAnswer};

    end

end



% If transformAllChannels is true we will, after the registration, run the 
% transformation all available channels for the sample2ARA data.
if transformAllChannels && length(nonRegChannels)>1
    nonRegChannels = nonRegChannels(~strcmp(dsFile,nonRegChannels));
else
    nonRegChannels = [];
end 


fprintf('\nRunning registration on volume %s\n\n',dsFile)


templateFile = getARAfnames;
if isempty(templateFile)
    return  %warning message already issued
end

%The path to the sample file
sampleFile = fullfile(downsampleDir,dsFile);
if ~exist(sampleFile,'file')
    fprintf('Can not find sample file at %s\n', sampleFile), return
end


%load the images
fprintf('Loading image volumes...')
templateVol = mhd_read(templateFile);
[~,~,ext] = fileparts(sampleFile);
switch ext
    case '.mhd'
        sampleVol = mhd_read(sampleFile);
    case '.tif'
        sampleVol = aratools.loadTiffStack(sampleFile);
end


% If median filtering was requested, filter the sample image using a filter
% of the desired width.
if medFiltSample
    origVol = sampleVol;
    sampleVol = medfilt3(sampleVol, repmat(medFiltSize,1,3));
end


% Log to a file the registration parameters
logFname = fullfile(regDir,'registration_log.txt');
logRegInfoToFile(logFname,sprintf('Analysis carried out: %s\n', datestr(now,'yyyy-mm-dd_HH-MM-SS')),true)
logRegInfoToFile(logFname,sprintf('Directory: %s\n', regDir))
logRegInfoToFile(logFname,sprintf('Voxel size: %d microns\n', S.ARAsize))
logRegInfoToFile(logFname,sprintf('Sample volume file: %s\n', sampleFile))
logRegInfoToFile(logFname,sprintf('Template file: %s\n', templateFile))

elastix_version=elastix('version');
logRegInfoToFile(logFname,sprintf('Registration software: %s\n', elastix_version))
if ~medFiltSample
    logRegInfoToFile(logFname,sprintf('Filtering of sample volume: none\n'))
else
    logRegInfoToFile(logFname,sprintf('Filtering of sample volume: medfilt3 with filter size %d\n',medFiltSize))
end

logRegInfoToFile(logFname,sprintf('\n\n**************  LOGGING  **************\n\n'))

%We should now be able to proceed with the registration. 
elastixDir = {};
if ara2sample
    %make the directory in which we will conduct the registration
    elastixDir{end+1} = fullfile(regDir,S.ara2sampleDir);
    if ~mkdir(elastixDir{end})
        fprintf('Failed to make directory %s',elastixDir{end})
    else

        writeLoggingLine(logFname,sprintf('Beginning registration of ARA to sample\n'))
        writeLoggingLine(logFname,sprintf('Conducting registration in %s\n',elastixDir{end}))
        [~,params]  = elastix(templateVol,sampleVol,elastixDir{end},-1,'paramstruct',elastixParams);
        writeLoggingLine(logFname,sprintf('Finished registration of ARA to sample\n'))

        if ~iscell(params.TransformParameters)
            msg=sprintf('\n\n\t** Transforming the ARA to the sample failed (see above).\n\t** Check Elastix parameters and your sample volumes\n\n');
            writeLoggingLine(logFname,msg)
        else
            if medFiltSample
                nonFiltFname = fullfile(elastixDir{end},'result.tiff');
                writeLoggingLine(logFname,sprintf('Transforming the original dataset to generate a non-filtered image: %s\n',nonFiltFname))
                RES = transformix(origVol,elastixDir{end});
                save3Dtiff(RES,nonFiltFname);
            end
        end

        %optionally remove files used to conduct registration 
        if S.removeMovingAndFixed
            delete(fullfile(elastixDir{end},[S.ara2sampleDir,'_moving*']))
            delete(fullfile(elastixDir{end},[S.ara2sampleDir,'_target*']))
        end
    end

end


if sample2ara
    %make the directory in which we will conduct the registration
    elastixDir{end+1} = fullfile(regDir,S.sample2araDir);
    if ~mkdir(elastixDir{end})
        writeLoggingLine(logFname,sprintf('Failed to make directory %s\n',elastixDir{end}))

    else
        writeLoggingLine(logFname,sprintf('Beginning registration of sample to ARA\n'))
        writeLoggingLine(logFname,sprintf('Conducting registration in %s\n',elastixDir{end}))
        [~,params]=elastix(sampleVol,templateVol,elastixDir{end},-1,'paramstruct',elastixParams); 
        writeLoggingLine(logFname,sprintf('Finished registration of sample to ARA\n'))

        if ~iscell(params.TransformParameters)
            writeLoggingLine(logFname,sprintf('\n\n\t** Transforming the sample to the ARA failed (see above).\n\t** Check Elastix parameters and your sample volumes\n'))
            writeLoggingLine(logFname,sprintf('\t** Not initiating inverse transform.\n\n'))
            return
        else
            if medFiltSample
                nonFiltFname = fullfile(elastixDir{end},'result.tiff');
                writeLoggingLine(logFname,sprintf('Transforming the original dataset to generate a non-filtered image: %s\n',nonFiltFname))
                RES = transformix(origVol,elastixDir{end});
                save3Dtiff(RES,nonFiltFname);
            end
    end

    if transformAllChannels
        fprintf('\n\nTransforming non-registered channels to ARA space using registrations in %s\n',elastixDir{end})
        for ii = 1:length(nonRegChannels)
            tFile = fullfile(downsampleDir,nonRegChannels{ii});
            aratools.applyRegToStackFile(tFile,elastixDir{end});
        end
    end %transformAllChannels

end % sample2ara


if ~suppressInvertSample2ara
        writeLoggingLine(logFname,sprintf('Beginning inversion of sample to ARA\n'))
        inverted=invertElastixTransform(elastixDir{end});
        save(fullfile(elastixDir{end},S.invertedMatName),'inverted')

        %Now we can transform the sparse points files 
        invertExportedSparseFiles(inverted)
        writeLoggingLine(logFname,sprintf('Finished inversion of sample to ARA\n'))
    end
    if S.removeMovingAndFixed
        delete(fullfile(elastixDir{end},[S.sample2araDir,'_moving*']))
        delete(fullfile(elastixDir{end},[S.sample2araDir,'_target*']))
    end
end %suppressInvertSample2ara

logRegInfoToFile(logFname,sprintf('\n\nCompleted at: %s\n', datestr(now,'yyyy-mm-dd_HH-MM-SS')))

% Copy log files into all reg-sub-dirs
cellfun(@(x) copyfile(logFname,x), elastixDir)

fprintf('\nFinished\n')




function logRegInfoToFile(fname,dataToLog,flush,echo)
    %Write string dataToLog to fname.
    %This little function is just to make it easier to log identity of the channel being registered
    %
    % fname - file to which we will log stuff
    % dataToLog - text to write
    % flush - optional. False if missing or empty. If true, we wipe file "fname" and start over.
    %         i.e. if flush is false we append.
    % echo - true by default. copies to screen the line written to file

    if nargin<3 || isempty(flush)
        flush=false;
    end

    if nargin<4
        echo=true;
    end

    if flush
        fid = fopen(fname,'w+');
    else
        fid = fopen(fname,'a+');
    end

    if fid<0
        fprintf('FAILED TO WRITE LOG DATA TO FILE %s\n',fname)
        return
    end

    fprintf(fid,dataToLog);

    fclose(fid);

    if echo
        fprintf(dataToLog)
    end


function writeLoggingLine(fname,messageToLog)
    % Function writes a line to the logging file, fname, in the format:
    % yyy-mm-dd_HH:MM:SS - INFO - mfilename - message
    % echos to screen
    % 
    % messageToLog - a string. e.g. output of sprintf. It is chomped. 

    fid = fopen(fname,'a+');
    msg = sprintf('%s - INFO - %s - %s\n', ...
        datestr(now,'yyyy-mm-dd_HH-MM-SS'), ...
        mfilename, strip(messageToLog));
    fprintf(fid,msg);
    fclose(fid);

    fprintf(msg) % write to screen
